src/libraries/Common/src/System/Net/NetworkInformation/InterfaceInfoPal.Windows.cs

```diff
@@ -1,13 +1,102 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Runtime.InteropServices;
+using System.Text;
+
 namespace System.Net.NetworkInformation
 {
     internal static class InterfaceInfoPal
     {
+        private const int StackAllocationThreshold = 512;
+
         public static uint InterfaceNameToIndex(string interfaceName)
         {
             return Interop.IpHlpApi.if_nametoindex(interfaceName);
         }
+
+        public static unsafe uint InterfaceNameToIndex(ReadOnlySpan<char> interfaceName)
```

Comment:
This is a whole lot of logic just to support the ANSI conversion. Could we instead follow the description at:
https://learn.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-if_nametoindex

> The if_nametoindex function is implemented for portability of applications with Unix environments, but the ConvertInterface functions are preferred. The if_nametoindex function can be replaced by a call to the [ConvertInterfaceNameToLuidA](https://learn.microsoft.com/en-us/windows/desktop/api/netioapi/nf-netioapi-convertinterfacenametoluida) function to convert the ANSI interface name to a [NET_LUID](https://learn.microsoft.com/en-us/windows/desktop/api/ifdef/ns-ifdef-net_luid_lh) followed by a call to the [ConvertInterfaceLuidToIndex](https://learn.microsoft.com/en-us/windows/desktop/api/netioapi/nf-netioapi-convertinterfaceluidtoindex) to convert the NET_LUID to the local interface index.

but using `ConvertInterfaceNameToLuidW` instead of `ConvertInterfaceNameToLuidA`? That should make it pretty streamlined and we wouldn't need most of this?

Reply:
Doing so would definitely make the marshalling easier. It'd also improve the efficiency of parsing a `ReadOnlySpan<byte>` - at the moment, we have to perform a two conversions from UTF8 -> Unicode -> ANSI.

I can't find any reference to errata for if_nametoindex, and I've run a few simple checks on my laptop, which covers a fairly wide set of bases on one OS (physical adapter, unplugged physical adapter, removed device, tunnel adapter, miniport, loopback interface.)

ConvertInterfaceNameToLuid has an additional documented length constraint (albeit a very large one) on the interface name though, so I'm a little concerned that there might be other differences between the two which I've just not seen. I'd prefer to split the replacement into a separate PR to make a possible revert easier, unless you definitely want it in this one.

Reply:
Thanks. We're no longer in a rush, and I'd rather not add a lot of code we know is about to go away, so I'd be inclined to just have it in this one.

Reply:
I've swapped to the new methodology now. Doing so also exposed a lack of testing for the conversion from interface name to index, so I've added a new InterfaceInfoPalTests class to provide a bit more confidence that the marshalling works properly. I can see these tests passing in CI on Linux, Windows and Mac, so I'm happy enough with the results.